minimal_memory_lock/minimal_memory_lock.cpp

```diff
@@ -0,0 +1,273 @@
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  // Argument count and usage
+  if (rcutils_cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  // Configuration variables
+  size_t max_process_memory = 0;
+  size_t stack_size = 0;
+  bool lock_memory = false;
+  bool use_sleep = false;
+
+  // Optional argument parsing
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_LOCK_MEMORY)) {
+    lock_memory = true;
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_MEMORY_PREALLOCATION)) {
+    max_process_memory = 1024 * 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_MEMORY_PREALLOCATION));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_THREAD_STACK_SIZE)) {
+    stack_size = 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_THREAD_STACK_SIZE));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_USE_SLEEP)) {
+    use_sleep = true;
+  }
+
+  // Change a new default thread stack size to reduce memory foot-print
+  if (stack_size > 0) {
+    set_default_thread_stacksize(stack_size);
+  }
+
+  print_process_memory("Process memory before node creation: ");
+
+  // Create nodes
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<MinimalPublisher>();
+
+  // Sleep required to avoid some initial minor page faults
+  // TODO(carlos): find out why we see minor page faults initially
+  // This could be caused by middleware threads being created concurrently
+  if (use_sleep) {
+    std::cout << "Sleeping here" << std::endl;
+    std::this_thread::sleep_for(500ms);
+  }
+
+  if (lock_memory) {
+    print_process_memory("Process memory before locking: ");
+    bool memory_is_locked = configure_malloc_behavior();
+    if (memory_is_locked) {
+      std::cout << "Memory locked" << std::endl;
+    } else {
+      return EXIT_FAILURE;
+    }
+    print_process_memory("Process memory after locking: ");
+  }
+  if (max_process_memory > 0) {
+    std::cout << "Memory to reserve: " <<
+      std::to_string(max_process_memory / (1024 * 1024)) << " MB" << std::endl;
+    auto current_process_memory = get_process_memory() * 1024;
+    if (current_process_memory > max_process_memory) {
+      std::cout << "Process memory higher than maximum memory to reserve: " << std::endl;
+      return EXIT_FAILURE;
+    }
+    auto additional_memory_to_reserve = max_process_memory - current_process_memory;
+    bool memory_is_reserved = reserve_process_memory(additional_memory_to_reserve);
+    if (!memory_is_reserved) {
+      return EXIT_FAILURE;
+    }
+    print_process_memory("Process memory after memory pool reservation: ");
+  }
+
+  print_process_memory("Process memory before spin: ");
+  print_page_faults("Total page faults before spin");
+  get_new_page_faults();  // init page fault count
+
+  rclcpp::spin(node);
+
+  print_process_memory("Process memory after spin: ");
+  print_page_faults("Total page faults after spin");
```

Comment:
I think for the demo, you might consider reading and writing to the memory you allocated and then make sure there's no page fault there? This way you can prove that demand paging has been turned off. It's not a big deal tho.

I do this here: https://github.com/shuhaowu/rt-demo/blob/7116d52/docs/prefault-experiments/prefault-test.cc#L73-L77

Reply:
Yep, I will do that